Mathlib/RepresentationTheory/Coinvariants.lean

```diff
@@ -0,0 +1,182 @@
+`V` by this submodule. We show that the functor sending a representation to its coinvariants is
+left adjoint to the functor equipping a module with the trivial representation.
+
+-/
+
+universe u v
+
+namespace Representation
+
+variable {k G V W : Type*} [CommRing k] [Monoid G] [AddCommGroup V] [Module k V]
+  [AddCommGroup W] [Module k W] (ρ : Representation k G V) (τ : Representation k G W)
+
+/-- The submodule of a representation generated by elements of the form `ρ g x - x`. -/
+def augmentationSubmodule : Submodule k V :=
+  Submodule.span k (Set.range fun (x : G × V) => ρ x.1 x.2 - x.2)
+
+variable {ρ}
+
+lemma mem_augmentationSubmodule_of_eq (g : G) (x : V) (a : V) (h : ρ g x - x = a) :
+    a ∈ augmentationSubmodule ρ :=
+  Submodule.subset_span ⟨(g, x), h⟩
+
+variable (ρ)
+
+/-- The coinvariants of a representation, `V ⧸ ⟨{ρ g x - x | g ∈ G, x ∈ V}⟩`. -/
+def Coinvariants := V ⧸ augmentationSubmodule ρ
+
+namespace Coinvariants
+
+instance : AddCommGroup (Coinvariants ρ) := by unfold Coinvariants; infer_instance
+instance : Module k (Coinvariants ρ) := by unfold Coinvariants; infer_instance
+
+/-- The quotient map from a representation to its coinvariants as a linear map. -/
+def mk : V →ₗ[k] Coinvariants ρ := Submodule.mkQ (augmentationSubmodule ρ)
+
+theorem mk_eq_iff {x y : V} :
+    mk ρ x = mk ρ y ↔ x - y ∈ ρ.augmentationSubmodule :=
+  Submodule.Quotient.eq _
+
+@[simp]
+lemma mk_ρ_apply (g : G) (x : V) :
+    mk ρ (ρ g x) = mk ρ x :=
+  (mk_eq_iff _).2 <| mem_augmentationSubmodule_of_eq g x _ rfl
+
+@[elab_as_elim]
+theorem induction_on {C : Coinvariants ρ → Prop} (x : Coinvariants ρ) (h : ∀ v : V, C (mk ρ v)) :
+    C x :=
+  Submodule.Quotient.induction_on _ x h
+
+/-- A `G`-invariant linear map induces a linear map out of the coinvariants of a
+`G`-representation. -/
+def lift (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) :
+    ρ.Coinvariants →ₗ[k] W :=
+  Submodule.liftQ _ f <| Submodule.span_le.2 fun x ⟨⟨g, y⟩, hy⟩ => by
+    simpa only [← hy, SetLike.mem_coe, LinearMap.mem_ker, map_sub, sub_eq_zero, LinearMap.coe_comp,
+      Function.comp_apply] using LinearMap.ext_iff.1 (h g) y
+
+variable {ρ}
+
+@[simp]
+theorem lift_comp_mk (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) :
+    lift ρ f h ∘ₗ mk ρ = f := rfl
+
+@[simp]
+theorem lift_mk (f : V →ₗ[k] W) (h : ∀ (x : G), f ∘ₗ ρ x = f) (x : V) :
+    lift ρ f h (mk _ x) = f x := rfl
+
+@[ext]
+lemma hom_ext {f g : Coinvariants ρ →ₗ[k] W} (H : f ∘ₗ mk ρ = g ∘ₗ mk ρ) : f = g :=
+  Submodule.linearMap_qext _ H
+
+variable (ρ)
+
+/-- Given `G`-representations on `k`-modules `V, W`, a linear map `V →ₗ[k] W` commuting with
+the representations induces a `k`-linear map between the coinvariants. -/
+noncomputable def map (f : V →ₗ[k] W) (hf : ∀ g, f ∘ₗ ρ g = τ g ∘ₗ f) :
+    Coinvariants ρ →ₗ[k] Coinvariants τ :=
+  lift _ (mk _ ∘ₗ f) fun g => LinearMap.ext fun x => (mk_eq_iff _).2 <|
+    mem_augmentationSubmodule_of_eq (ρ := τ) g (f x) _ <| by simpa using congr($((hf g).symm) x)
```

Comment:
(could probably be golfed using the above ext lemma but big deal)

Reply:
(it's a map out of `V` so afraid not)